[#505] create an atomic ACLs endpoint (#808)
irods/python-irodsclient: README.md

`<session>.permissions` was therefore removed in v2.0.0
in favor of `<session>.acls`.

Atomic ACLs
-----------

A list of permissions may be added to an object atomically using
the AccessManager's apply_atomic_operations method:
```
from irods.access import ACLOperation
from irods.helpers import home_collection
session = irods.helpers.make_session()
myCollection = session.collections.create(f"{home_collection(session).path}/newCollection")

session.acls.apply_atomic_operations(myCollection.path,
   *[ACLOperation("read", "public"),
     ACLOperation("write", "bob", "otherZone")
    ])
```
ACLOperation objects form a linear order with iRODSAccess objects, and

> indeed are subclassed from them as well, allowing equivalency testing:

Reviewer: ```suggestion
ACLOperation objects form a linear order with iRODSAccess objects, and
indeed are subclassed from them as well, allowing equivalency testing.
```

Author: Consider it done.

`<session>.permissions` was therefore removed in v2.0.0
in favor of `<session>.acls`.

> Atomic ACLs

Reviewer 2: Consider including "Permissions" in the title. That will make the section easier to find for some readers.

Author: done

-----------

A list of permissions may be added to an object atomically using

> the AccessManager's apply_atomic_operations method:

Reviewer 2: ```suggestion
the AccessManager's `apply_atomic_operations` method:
```

> ```

Reviewer 2: ```suggestion
```py
```

from irods.access import ACLOperation
from irods.helpers import home_collection
session = irods.helpers.make_session()
myCollection = session.collections.create(f"{home_collection(session).path}/newCollection")

session.acls.apply_atomic_operations(myCollection.path,
   *[ACLOperation("read", "public"),
     ACLOperation("write", "bob", "otherZone")

>     ])

Reviewer 2: ```suggestion
session.acls.apply_atomic_operations(
    myCollection.path,
    *[
        ACLOperation("read", "public"),
        ACLOperation("write", "bob", "otherZone")
    ]
)
```

```

> ACLOperation objects form a linear order with iRODSAccess objects, and

Reviewer 2: ```suggestion
`ACLOperation` objects form a linear order with `iRODSAccess` objects, and
```

indeed are subclassed from them as well, allowing equivalency testing.

Thus, for example:

> ```

Reviewer 2: ```suggestion
```py
```

ACLOperation('read','public') in sess.acls.get(object)
```
is a valid operation, so that an application that tends to cache object
permissions client-side might use such checks in optimizing atomic ACL

> requests against the inclusion of any redundant ACLOperations.

Reviewer 2: ```suggestion
requests against the inclusion of any redundant ACL operations.
```

For purposes of sorting, a `__lt__` operator is also defined that would

> allow sorting of lists of iRODSAccess, ACLOperations, or the two intermixed:

Reviewer 2: ```suggestion
allow sorting of lists of `iRODSAccess`, `ACLOperations`, or the two intermixed:
```

> ```

Reviewer 2: ```suggestion
```py
```

> perms_list=[ACLOperation('read','bob'), iRODSAccess('read','/tempZone/home/alice','alice'), ]

Reviewer 2: ```suggestion
perms_list=[
    ACLOperation('read', 'bob'),
    iRODSAccess('read', '/tempZone/home/alice', 'alice')
]
```

print(sorted(perms_list))
```

and, as always, a sort key may be used for custom sorting; for example,
the following sorts the objects simply by ascending numerical value of the access:

> ```

Reviewer 2: ```suggestion
```py
```

perms = sorted([ACLOperation('read','bob'), ACLOperation('write', 'rods'), ACLOperation('read_object','alice')], 

>     key=lambda acl: iRODSAccess.codes[acl.access_name]

Reviewer 2: ```suggestion
perms = sorted(
    [
        ACLOperation('read', 'bob'),
        ACLOperation('write', 'rods'),
        ACLOperation('read_object', 'alice')
    ], 
    key=lambda acl: iRODSAccess.codes[acl.access_name]
```

session.acls.apply_atomic_operations(
    myCollection.path,
    *[
        ACLOperation("read", "public"),
        ACLOperation("write", "bob", "otherZone")
    ]
)
```
`ACLOperation` objects form a linear order with `iRODSAccess` objects, and
indeed are subclassed from them as well, allowing equivalence comparisons and
also permitting intermixed sequences to be sorted (using the `__lt__` method
if no sort `key` parameter is given).

Care should be taken however to normalize the objects before such comparisons
and sorting, and with connected uses of the `in` operator (which leverages `__eq__`).

The following code sorts the objects based on their lexical order starting with the
normalized `access_name`, which serves to group identical permissions together:
```py
from irods.access import *
import irods.helpers
acls = [
    iRODSAccess('read_object', '/tempZone/home/alice', 'bob', 'tempZone'),
    ACLOperation('write', 'rods'),
    ACLOperation('read', 'bob'),
]

session = irods.helpers.make_session()
N = lambda acl: acl.normalize(local_zone=session.zone)

print(N(acls[0]) == N(acls[2]))
acls.sort(key=N)

> print( N(iRODSAccess('read', '', 'bob')) in map(N, acls) )

Reviewer 2: ```suggestion
print(N(iRODSAccess('read', '', 'bob')) in map(N, acls))
```
